src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+         MONITOR_LOG (server_monitor,
+                      "immediately proceeding due to moreToCome");
+         continue;
+      }
+
+      /* ... or the server has just transitioned to Unknown due to a network
+       * error. */
+      if (_mongoc_error_is_network (&description->error) &&
+          previous_description->type != MONGOC_SERVER_UNKNOWN) {
+         MONITOR_LOG (server_monitor,
+                      "immediately proceeding due to network error");
+         continue;
+      }
+
+      mongoc_server_monitor_wait (server_monitor);
+   }
+
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   server_monitor->shared.state = MONGOC_THREAD_JOINABLE;
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+   mongoc_server_description_destroy (previous_description);
+   mongoc_server_description_destroy (description);
+   BSON_THREAD_RETURN;
+}
+
+static bool
+_server_monitor_ping_server (mongoc_server_monitor_t *server_monitor,
+                             int64_t *rtt_ms)
+{
+   bool ret = false;
+   int64_t start_us = _now_us ();
+   bson_t ismaster_reply;
+   bson_error_t error;
+
+   *rtt_ms = -1;
+
+   if (!server_monitor->stream) {
+      MONITOR_LOG (server_monitor, "rtt setting up connection");
+      ret = _server_monitor_setup_connection (
+         server_monitor, &ismaster_reply, &start_us, &error);
+      bson_destroy (&ismaster_reply);
+   }
+
+   if (server_monitor->stream) {
+      MONITOR_LOG (server_monitor, "rtt polling ismaster");
+      ret = _server_monitor_polling_ismaster (
+         server_monitor, &ismaster_reply, &error);
+      if (ret) {
+         *rtt_ms = (_now_us () - start_us) / 1000;
+      }
+      bson_destroy (&ismaster_reply);
+   }
+   return ret;
+}
+
+/* The RTT monitor thread function.
+ *
+ * Server monitor must be in state MONGOC_THREAD_OFF.
+ */
+static BSON_THREAD_FUN (_server_monitor_rtt_thread, server_monitor_void)
+{
+   mongoc_server_monitor_t *server_monitor;
+
+   server_monitor = (mongoc_server_monitor_t *) server_monitor_void;
+
+   while (true) {
+      int64_t rtt_ms;
+      bson_error_t error;
+
+      bson_mutex_lock (&server_monitor->shared.mutex);
+      if (server_monitor->shared.state != MONGOC_THREAD_RUNNING) {
+         bson_mutex_unlock (&server_monitor->shared.mutex);
+         break;
+      }
+      bson_mutex_unlock (&server_monitor->shared.mutex);
+
+      _server_monitor_ping_server (server_monitor, &rtt_ms);
+      if (rtt_ms != -1) {
+         mongoc_server_description_t *sd;
+
+         bson_mutex_lock (&server_monitor->topology->mutex);
+         sd = mongoc_topology_description_server_by_id (
+            &server_monitor->topology->description,
+            server_monitor->description->id,
+            &error);
+         if (sd) {
+            /* If the server description has been removed, the RTT thread will
+             * be terminated by background monitoring soon. */
+            mongoc_server_description_update_rtt (sd, rtt_ms);
```

Comment:
Should `sd` be destroyed before we release the topology lock, or is this returning an `sd` directly from the topology's set, which is why we hold the topology lock while working with it?